### PR TITLE
dummy file location change

### LINKED ISSUE #6471: IAM minimal permissions: dummy file dependency fails creating repository
**Background issue:** #6299 

**Context:** 

- We want to work in pre-sign mode in AWS S3 and block lakeFS from Get/Put to the bucket, ideally only allow `*/_lakefs/*` folder. 
- When configuring in AWS S3 the desired permissions we are not able to create a repository because lakeFS creates `dummy` file in the root of the storage namespace. 

**D.O.D:**  

- We can create lakeFS repository while lakeFS can only work with `_lakefs/*` directory or some other scoped directory. 
- Potentially for simplicity in such a case we can even simply add configuration to **skip**  dummy file check and everything that's related to it.

### PR DESCRIPTION
**Pull request by Isan-Rivkin**
Closes #6471 

**Solution**
- Create `dummy` file in `_lakefs/dummy` 
- Maintains backward compatibility (repositories created prior to this version will be checked to with `dummy` at the root namespace) 
- Will break if downgraded and repositories created in this version.
- Side Effect: Removed support for migration from lakeFS version < v0.50. It's already outdated and otherwise the dummy file raises more questions. (i.e what happens if skipping accessibility test when repo created on Azure).

### CODE REVIEW

File: docs/howto/deploy/upgrade.md [hunk 1]
```diff
@@ -136,86 +136,3 @@ To do so, add the following to your configuration (relevant **only** for 0.30.0)
 cataloger:
   type: rocks
 ```
-
```
Comment (arielshaqed) [inline docs/howto/deploy/upgrade.md:139]: Can we instead add a note along these lines?

```suggestion
If you are using a version before 0.50.0, you must first perform the [previous upgrade to that version](https://docs.lakefs.io/v0.50/reference/upgrade.html#data-migration-for-version-v0500). {: note: .note-warning }

File: pkg/api/controller.go [hunk 2]
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
 }
 
 func (c *Controller) ensureStorageNamespace(ctx context.Context, storageNamespace string) error {
-	const (
-		dummyKey  = "dummy"
-		dummyData = "this is dummy data - created by lakeFS in order to check accessibility"
-	)
+	dummyKey := fmt.Sprintf("%s/dummy", c.Config.Committed.BlockStoragePrefix)
+	const dummyData = "this is dummy data - created by lakeFS in order to check accessibility"
```
Comment (arielshaqed) [inline pkg/api/controller.go:1630]: If we're making changes...
```suggestion
	const dummyData = "this is dummy data - created by lakeFS to check accessibility"
```

File: pkg/api/controller.go [hunk 3]
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
...
 
-	obj := block.ObjectPointer{
+	// check if dummy file exists in previous lakeFS versions at the root of the storage namespace
+	objLegacyLocation := block.ObjectPointer{
 		StorageNamespace: storageNamespace,
 		IdentifierType:   block.IdentifierTypeRelative,
-		Identifier:       dummyKey,
+		Identifier:       "dummy",
```
Comment (arielshaqed) [inline pkg/api/controller.go:1636]: Suggest keeping this as a key `dummyLegacyKey` or somesuch.  If you do that you probably don't need the comment at l.1632.

File: pkg/api/controller.go [hunk 4]
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
...
 	}
 	objLen := int64(len(dummyData))
-	if _, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
-		return fmt.Errorf("found lakeFS objects in the storage namespace(%s): %w",
+
+	if _, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
```
Comment (arielshaqed) [inline pkg/api/controller.go:1640]: YIKES!  As Mitch Hedberg would say,
> _I used to leak a file descriptor.  I still do.  But I used to, too.

Consider using `.Exists()` (which doesn't test readability), or just closing (which probably also tests readability):
```suggestion
	if exists, err := c.BlockAdapter.Exists(ctx, objLegacyLocation, objLen); err == nil {
```
or
```suggestion
	if s, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
		s.Close()
```

File: pkg/api/controller.go [hunk 5]
```diff
@@ -1626,23 +1626,36 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
...
 		return err
 	}
+	// check if the dummy file exist
+	obj := block.ObjectPointer{
+		StorageNamespace: storageNamespace,
+		IdentifierType:   block.IdentifierTypeRelative,
+		Identifier:       dummyKey,
+	}
+
+	if _, err := c.BlockAdapter.Get(ctx, objLegacyLocation, objLen); err == nil {
```
Comment (arielshaqed) [inline pkg/api/controller.go:1653]: Obviously the same applies here, too.

Comment (arielshaqed) [inline pkg/api/controller.go:1653]: Bug?
```suggestion
	if _, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
```

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Neat!  Some minor but important requests.

**Comment by arielshaqed** [comment]
> * Side Effect: Removed support for migration from lakeFS version < v0.50. It's already outdated and otherwise the dummy file raises more questions. (i.e what happens if skipping accessibility test when repo created on Azure).

This is important for the CHANGELOG.  Can you add a

File: pkg/api/controller.go (hunk 5 above)
Comment (arielshaqed) [reply pkg/api/controller.go:1653]: Probably merits a test in controller_test.go.  The important part is to test that re-creating another repo at the same location fails; less important is to verify that creating another repo at the same location as a "legacy" dummy repo also fails.

File: docs/howto/deploy/upgrade.md (hunk 1 above)
Comment (Isan-Rivkin) [reply docs/howto/deploy/upgrade.md:139]: 👍

File: pkg/api/controller.go (hunk 2 above)
Comment (Isan-Rivkin) [reply pkg/api/controller.go:1630]: breaking backward compatibility with transient storage but we can live with that 😃

File: pkg/api/controller.go (hunk 5 above)
Comment (Isan-Rivkin) [reply pkg/api/controller.go]: isn't it what [use same storage namespace twice](https://github.com/treeverse/lakeFS/blob/master/pkg/api/controller_test.go#L203) test does?

Comment (Isan-Rivkin) [reply pkg/api/controller.go]: Changed implementation please look

File: pkg/api/controller.go (hunk 4 above)
Comment (Isan-Rivkin) [reply pkg/api/controller.go]: Good 👁️ you have!

File: pkg/api/controller.go (hunk 3 above)
Comment (Isan-Rivkin) [reply pkg/api/controller.go]: Changed implementation please look

File: Makefile [hunk 6]
```diff
@@ -165,6 +165,10 @@ build: gen docs ## Download dependencies and build the default binary
 	$(GOBUILD) -o $(LAKEFS_BINARY_NAME) -ldflags $(LD_FLAGS) -v ./cmd/$(LAKEFS_BINARY_NAME)
 	$(GOBUILD) -o $(LAKECTL_BINARY_NAME) -ldflags $(LD_FLAGS) -v ./cmd/$(LAKECTL_BINARY_NAME)
 
+# TODO(isan) delete before merge
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply Makefile]: 👍

**Comment by Isan-Rivkin** [comment]
@arielshaqed please check again, I addressed all of your great comments + changed the implementation. 

Regarding "Legacy": 
Before recent commit: I did legacy check if `dummy` or `_lakefs/dummy` exists, now I changed it to only check `_lakefs/dummy`. 
The reason is that when clients will use minimal IAM permissions they will not be able to `Get` object `dummy`, they will get permission deny and that creates weird behavior questions in lakeFS - that can be addressed by adding more configuration but I'd rather not keeping this backward compatible for simplicity - @ozkatz please verify that's what we discussed.

File: pkg/config/config.go [hunk 7]
```diff
@@ -228,9 +228,10 @@ type Config struct {
 		} `mapstructure:"ui_config"`
 	} `mapstructure:"auth"`
 	Blockstore struct {
-		Type                   string  `mapstructure:"type" validate:"required"`
-		DefaultNamespacePrefix *string `mapstructure:"default_namespace_prefix"`
-		Local                  *struct {
+		Type                        string  `mapstructure:"type" validate:"required"`
+		DefaultNamespacePrefix      *string `mapstructure:"default_namespace_prefix"`
+		EnsureRootNamespaceRWAccess bool    `mapstructure:"ensure_root_namespace_rw_access"`
```
Comment (arielshaqed) [inline pkg/config/config.go:233]: Yay, naming!  AFAICT this only causes read-access to be verified, and not even that (if you have just new repos).

Consider
```suggestion
		ReadableRootNamespace bool    `mapstructure:"readable_root_namespace"`
```

Also arguably this is better a part of Graveler than of Blockstore: it has no influence on blockstore behaviour.

File: pkg/config/defaults.go [hunk 8]
```diff
@@ -60,6 +60,7 @@ func setDefaults(cfgType string) {
 	viper.SetDefault("auth.remote_authenticator.default_user_group", "Viewers")
 	viper.SetDefault("auth.remote_authenticator.request_timeout", 10*time.Second)
 
+	viper.SetDefault("blockstore.ensure_root_namespace_rw_access", true)
```
Comment (arielshaqed) [inline pkg/config/defaults.go:63]: And also update docs/reference/configuration.md, please :-/

**Review by arielshaqed — APPROVED** [review]
Thanks!

Approved to save time, but I think we should improve the config option name, and we **must** document it.

File: pkg/config/config.go (hunk 7 above)
Comment (arielshaqed) [reply pkg/config/config.go:233]: @nopcoder probably can say.

File: pkg/api/controller.go [hunk 9]
```diff
@@ -1626,20 +1626,21 @@ func (c *Controller) validateStorageNamespace(storageNamespace string) error {
...
 		IdentifierType:   block.IdentifierTypeRelative,
 		Identifier:       dummyKey,
 	}
 	objLen := int64(len(dummyData))
-	if _, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
-		return fmt.Errorf("found lakeFS objects in the storage namespace(%s): %w",
-			storageNamespace, ErrStorageNamespaceInUse)
+
+	if s, err := c.BlockAdapter.Get(ctx, obj, objLen); err == nil {
+		defer s.Close()
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/api/controller.go:1664]: 👍

File: pkg/config/defaults.go (hunk 8 above)
Comment (Isan-Rivkin) [reply pkg/config/defaults.go:63]: done

**Pull request merged**